{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "runs {'A': 0, 'C': 2, 'G': 0, 'U': 3}\n",
      "0.5512437810945273\n",
      "runs {'A': 0, 'C': 0, 'G': 0, 'U': 0}\n",
      "0.5545545545545546\n"
     ]
    }
   ],
   "source": [
    "file = open('/Users/na395/Desktop/Python files/sequential_prot_mt.txt','r')\n",
    "import re\n",
    "protseq = file.readline().rstrip('\\n')\n",
    "\n",
    "file.close()\n",
    "\n",
    "file = open('/Users/na395/Desktop/Python files/prot_seq_and_no_run_dna.txt','w')\n",
    "\n",
    "#is able to count how many 3> runs of nts in total there are in seq\n",
    "def number_of_runs(seq):\n",
    "    total = (0)\n",
    "    for each in ('ACGU'):\n",
    "        total = total + len(re.findall(each + '{4,}', seq))\n",
    "    return(total)\n",
    "\n",
    "#returns GC percent content in seq\n",
    "def GC_content(seq):\n",
    "    GCcount=((seq.count('G')+seq.count('C'))/len(seq))\n",
    "    return(GCcount)\n",
    "\n",
    "#codondicts is a list of dicts (7), dict 0 = how many codons each aa has, then dict 1 = most used codon, dict 2 = 2nd etc. \n",
    "#get value by codondicts[x][key] where x is desired dict\n",
    "#have manually swapped D, C, F codons around as is only 53 vs 47 % useage and will help gc content. P is 32 vs 31%.S and R codons also swapped though % the same\n",
    "\n",
    "codondicts=[{'A': '4', 'C': '2', 'D': '2', 'E': '2', 'F': '2', 'G': '4', 'H': '2', 'I': '3', 'K': '2', 'L': '6', 'M': '1', 'N': '2', 'P': '4', 'Q': '2', 'R': '6', 'S': '6', 'T': '4', 'V': '4', 'W': '1', 'Y': '2'}\n",
    ",{'A': 'GCC', 'C': 'UGU', 'D': 'GAU', 'E': 'GAG', 'F': 'UUU', 'G': 'GGC', 'H': 'CAC', 'I': 'AUC', 'K': 'AAG', 'L': 'CUG', 'M': 'AUG', 'N': 'AAC', 'P': 'CCU', 'Q': 'CAG', 'R': 'AGA', 'S': 'UCU', 'T': 'ACC', 'V': 'GUG', 'W': 'UGG', 'Y': 'UAC'}\n",
    ",{'A': 'GCU', 'C': 'UGC', 'D': 'GAC', 'E': 'GAA', 'F': 'UUC', 'G': 'GGA', 'H': 'CAU', 'I': 'AUU', 'K': 'AAA', 'L': 'CUC', 'N': 'AAU', 'P': 'CCC', 'Q': 'CAA', 'R': 'AGG', 'S': 'UCC', 'T': 'ACA', 'V': 'GUC', 'Y': 'UAU'}\n",
    ",{'A': 'GCA', 'G': 'GGG', 'I': 'AUA', 'L': 'UUG', 'P': 'CCA', 'R': 'CGG', 'S': 'AGC', 'T': 'ACU', 'V': 'GUU'}\n",
    ",{'A': 'GCG', 'G': 'GGU', 'L': 'CUU', 'P': 'CCG', 'R': 'CGC', 'S': 'AGU', 'T': 'ACG', 'V': 'GUA'}\n",
    ",{'L': 'CUA', 'R': 'CGA', 'S': 'UCA'}\n",
    ",{'L': 'UUA', 'R': 'CGU', 'S': 'UCG'}]\n",
    "\n",
    "dnaseq = ['NNN']\n",
    "\n",
    "for each in protseq:\n",
    "    dnaseq.append(codondicts[1][each])\n",
    "dnaseq.append('NNN')\n",
    "dnaseqcont= (''.join(dnaseq))\n",
    "\n",
    "\n",
    "\n",
    "rundictnumbers={}\n",
    "for each in ('ACGU'):\n",
    "     rundictnumbers[each]= len(re.findall(each + '{4,}', dnaseqcont))   \n",
    "print('runs '+ str(rundictnumbers))    \n",
    "print(GC_content(dnaseqcont))\n",
    "\n",
    "\n",
    "rundict={}\n",
    "for each in ('ACGU'):\n",
    "    rundict[each]= re.finditer(each + '{4,}', dnaseqcont)\n",
    "    \n",
    "\n",
    "#gets start and end posn of match and gets codon numbers. Switches first codon to second dict and evaluates if the\n",
    "#new combo of codons (+1 at either side) still contains the run. If yes then moves to second codon and tries the \n",
    "#same. And then tries the third etc. If this fails mts first and second and cycles again. \n",
    "# If it runs out of loops then it means it cant be solved or flanking codons come into play.\n",
    "#This script will not trial combinations of three mutations\n",
    "\n",
    "for base in 'ACGU':\n",
    "\n",
    "    for all in rundict[base]:\n",
    "\n",
    "        firstcodon=((all.start()-all.start()%3)/3)\n",
    "        lastcodon=(((all.end()-1)-(all.end()-1)%3)/3) #need to +1 if you want this to be part of a range\n",
    "\n",
    "\n",
    "\n",
    "        tempcodons=[] #need tempcodons because it can be used to permanently change a codon and then trial changes of the others\n",
    "        tempcodons=dnaseq[int(firstcodon)-1:int(lastcodon)+2]\n",
    "        temptempcodons=tempcodons   #need temptempcodons because it can be used to try single changes of codons\n",
    "\n",
    "        for z in '123456': #23456 because that is how many codon dictionaries there are\n",
    "            if number_of_runs(''.join(temptempcodons)) == 0: #this is to stop the perm loop once the temp loop \n",
    "                break\n",
    "            else:\n",
    "                w=(0) #for marking which codon we are at\n",
    "                tempcodons=dnaseq[int(firstcodon)-1:int(lastcodon)+2]\n",
    "                for every in protseq[int(firstcodon)-1:int(lastcodon)]:\n",
    "                    w = int(w)+1\n",
    "                    if int(z)-1 >= int(codondicts[0][every]):\n",
    "                        #it means that all the permanent codons for this amino acid have been tried \n",
    "                        continue\n",
    "                    else:\n",
    "                        #set that codon to perm and then assess\n",
    "                        newpermcodon= (codondicts[int(z)][every])\n",
    "                        tempcodons[w]=newpermcodon\n",
    "                        for x in '123456': #23456 because that is how many codon dictionaries there are\n",
    "                                if number_of_runs(''.join(temptempcodons)) == 0:\n",
    "                                    break\n",
    "                                y=(0)         #for marking which codon we are at\n",
    "\n",
    "                                for each in protseq[int(firstcodon)-1:int(lastcodon)]:\n",
    "                                    y=int(y)+1\n",
    "                                    if int(x)-1 >= int(codondicts[0][each]):                            \n",
    "                                        #it means that all the temporary codons for this amino acid have been tried \n",
    "                                        continue\n",
    "                                    else:\n",
    "                                        temptempcodons=tempcodons #this resets temptempcodons to the original\n",
    "                                        newcodon = (codondicts[int(x)][each])\n",
    "                                        temptempcodons[y]=newcodon\n",
    "                                        if number_of_runs(''.join(temptempcodons)) == 0:\n",
    "                                            break\n",
    "\n",
    "                                        else:                             \n",
    "\n",
    "                                            pass\n",
    "                                            #it  means that the run of nts has NOT been removed\n",
    "                                            #it means that further codon mts need to be attempted\n",
    "\n",
    "\n",
    "        #hopefully temptempcodons can now be used to replace the original dnaseq\n",
    "\n",
    "\n",
    "\n",
    "        if number_of_runs(''.join(temptempcodons)) == 0:\n",
    "            #take these mutations and enter into original sequence\n",
    "            for x in range(len(temptempcodons)-2):\n",
    "                dnaseq[int(firstcodon)+int(x)]=temptempcodons[int(x)+1]\n",
    "\n",
    "\n",
    "                \n",
    "                \n",
    "                \n",
    "dnaseq.remove('NNN')\n",
    "dnaseq.remove('NNN')               \n",
    "\n",
    "dnaseqcont= (''.join(dnaseq))\n",
    "rundictnumbers={}\n",
    "for each in ('ACGU'):\n",
    "    rundictnumbers[each]= len(re.findall(each + '{4,}', dnaseqcont))\n",
    "\n",
    "print('runs '+ str(rundictnumbers))      \n",
    "print(GC_content(dnaseqcont))\n",
    "\n",
    "file.write(dnaseqcont.strip('N')+'\\n')\n",
    "\n",
    "file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "##### # GC content 40-60\n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "#realised that of the most common codons only proline causes C problems as it is CCC. All the other common codons \n",
    "#cannot cause >3 Cs\n",
    "#note WG will always have G{4}\n",
    "#\n",
    "# no 2* structure\n",
    "# should not encode certain r.e.'s\n",
    "# don't add or remove splice sites, check using gene splicer? keep mutagenesis away from edges...\n",
    "#    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
